Feature/v4 migration guide (#869)
MapsPeople/docs: src/content/getting-started/v4-migration-guide.md

---
title: Android BETA - Migrating from V3 to V4
toc: true
eleventyNavigation:
  key: getting-started-v4-migration-android
  title: Android BETA - Migrating from V3 to V4
---

> The Android SDK for MapsIndoors has been upgraded from V3 to V4, which comes with a lot of improvements to stability and flexibility. SDK loading has been improved, and is less error-prone. This guide will cover specific changes to the SDK and how to use it, to provide you with a guide on how to upgrade from V3 to V4.

Reviewer: Improvements to stability and less error prone, should be internal only IMO. 

Lets stay to what the V4 brings for convenience and features. We now have the possibility to use MapBox with MapsIndoors as well as multiple interfaces have been refactored to simplify development and SDK behaviour when developing as well as removing redundant boiler plate code

> The Android SDK for MapsIndoors has been upgraded from V3 to V4, which comes with a lot of improvements features and flexibility. The MapsIndoors SDK now supports Mapbox as a map provider, alongside containing some reworked and refactored features that simplify development and SDK behavior. This guide will cover specific changes to the SDK and how to use it, to provide you with a guide on how to upgrade from V3 to V4.

Reviewer: improvements, features. 

which comes with improved interfaces and flexibility for developing your own map experience.
@fhansenMP @matiasholst Thoughts?

Reviewer 2: yes

> The Android SDK for MapsIndoors has been upgraded from V3 to V4, which comes with a lot of improvements features and flexibility. The MapsIndoors SDK now supports Mapbox as a map provider, alongside containing some reworked and refactored features that simplify development and SDK behavior. This guide will cover specific changes to the SDK and how to use it, to provide you with a guide on how to upgrade from V3 to V4.

Reviewer 3: ```suggestion
The Android SDK for MapsIndoors has been upgraded from V3 to V4, which comes with a lot of improvements, features and flexibility. The MapsIndoors SDK now supports Mapbox as a map provider, alongside some reworked and refactored features that simplify development and SDK behavior. This guide will cover specific changes to the SDK and how to use it to provide you with a guide on how to upgrade from V3 to V4.
```

> ## MapsIndoors SDK Map Engine Flavours

Reviewer 3: ```suggestion
## MapsIndoors SDK Map Engine Flavors
```

> <!--- This part is not 100% settled, be sure to talk to Frederik to confirm before release!

Reviewer: With the release of V4 the MapsIndoors SDK is released as two separate libraries depending on the Map Provider.
You can get them through Maven by changing your dependency to get:

```java
implementation 'com.mapspeople.mapsindoors:googlemapssdk:4.0.0-beta.9'

implementation 'com.mapspeople.mapsindoors:mapboxsdk:4.0.0-beta.9'
´´´

As of V4 the MapsIndoors mobile SDKs support both Google Maps and Mapbox. A version of the SDK is available for each map engine.

MapsIndoorsSDKGoogleMaps 4.x.x

MapsIndoorsSDKMapbox 4.x.x

> --->

Reviewer 3: This should not be commented out since both map engines are available.

> Both of these are distributed through Maven, and come packaged with a number of helper classes.

Reviewer: This is slightly redundant by the line above now i guess

## MapsIndoors Initialization

MapsIndoors is a singleton class, which can be described as the data layer of the SDK. Below you will find an example that demonstrates how initialization has been simplified between V3 and V4.

### V3

In V3, SDK initialization is started with:

```java
MapsIndoors.initialize(getApplicationContext(), "mapsindoors-key", listener);
```

And subsequently setting the Google API key using:

```java
MapsIndoors.setGoogleAPIKey(getString(R.string.google_maps_key));
```

If you want to change the MapsIndoors API key of an already initialized SDK you invoke:

```java
MapsIndoors.setApiKey("new key")
```

And to close down the SDK, call:

```java
MapsIndoors.onApplicationTerminate()
```

### V4

> In this new version of the SDK, initialization is started by the new function `MapsIndoors.load()`:

Reviewer 3: ```suggestion
In V4, initialization is started by the new function `MapsIndoors.load()`:
```

```java
MapsIndoors.load(getApplicationContext(), "mapsindoors-key", listener);
```

> Map-engine-specific API keys are handled by `MPMapConfig`, covered in the "MapControl Initialization" section of this guide.

Reviewer 3: ```suggestion
Map engine specific API keys are handled by `MPMapConfig`, covered in the "MapControl Initialization" section of this guide.
```

Switching to another MapsIndoors API key, such as for switching active Solutions, is now done by invoking `MapsIndoors.load()` again with a new key. The SDK will close down, and reload with the new API key.

To close down the SDK without reloading a new API key, invoke:

```java
MapsIndoors.destroy()
```

## MapControl Initialization

> MapControl instantiation and initialization are separate concepts. You create a new instance of `MapControl` and configure it with a map and view - optionally you could set clustering, overlapping or other behavior on the object.

Reviewer 3: ```suggestion
MapControl instantiation and initialization are separate concepts. You create a new instance of `MapControl` and configure it with a map and view - optionally you could set clustering, overlapping and other behavior on the object.
```

### V3

> In V3, `MapControl` is a separate asynchronous call:

Reviewer 3: ```suggestion
In V3, `MapControl.init()` is a separate asynchronous call:
```

```java
mMapControl = new MapControl(this);
mMapControl.setGoogleMap(mMap, view);
mMapControl.init(miError -> {
    // MapControl init complete
});
```

### V4

In V4, `MapControl` now requires a `MPMapConfig` object, which is acquired using a builder on the class `MPMapConfig`. Here you must provide an activity, a map provider (Google Maps or Mapbox), a `mapview` and a map engine API key.

```java
MPMapConfig mapConfig = new MPMapConfig.Builder(activity, googleMap, "google-api-key", view, true)
        .setClusteringEnabled(false)
        .setAllowMarkerOverlap(false)
        .build();
```

> With a `MPMapConfig` instance, you may create a new `MapControl` instance. This now happens through a factory pattern. This both instantiates and initializes your `MapControl` object asynchronously. If everything succeeded, you will receive a ready-to-use `MapControl` instance - if not, you will get an error and receive no `MapControl` instance.

Reviewer 3: ```suggestion
With a `MPMapConfig` instance, you may create a new `MapControl` instance. This now happens through a factory pattern. This both instantiates and initializes your `MapControl` object asynchronously. If everything succeed, you will receive a ready-to-use `MapControl` instance - if not, you will get an error and receive no `MapControl` instance.
```

```java
MapControl.create(mapConfig, (mapControl, miError) -> {
    // MapControl init complete
});
```

Please note that this factory method will wait to return until a valid MapsIndoors solution is loaded, therefore it is safe to invoke `MapControl.create()` prior to, or in parallel with `MapsIndoors.load()`.

## SolutionConfig & AppConfig

### Clustering & Marker Collisions

#### V3

In V3, AppConfig contained information about clustering (`POI_GROUPING`) and collisions (`POI_HIDE_ON_OVERLAP`), which could be fetched and updated like this:

```java
// get whether collisions are enabled... as a string
MapsIndoors.getAppConfig().getAppSettings().get(AppConfig.APP_SETTING_POI_HIDE_ON_OVERLAP);

// set whether clustering is enabled... with a string
MapsIndoors.getAppConfig().getAppSettings().put(AppConfig.APP_SETTING_POI_GROUPING, "true");
```

#### V4

> In V4, these settings have been moved to `MPSolutionConfig`, which is located on the MPSolution, in this version these settings have [something about boolean and enum]. They can be fetched and updates like this:

Reviewer: something about boolean and enum?

Author: Matias will fix this 😄

> In V4, these settings have been moved to `MPSolutionConfig`, which is located on the MPSolution, in this version these settings have types (a boolean and an Enum type). This helps ensure that these settings are easier to configure and have no parsing errors. They can be fetched and updates like this:

Reviewer 3: ```suggestion
In V4, these settings have been moved to `MPSolutionConfig`, which is located on the MPSolution. Now these settings have types (a boolean and an Enum type). This helps ensure that the settings are easier to configure and have no parsing errors. They can be fetched and updates like this:
```

```java
// get the config from the solution
MPSolutionConfig config = MapsIndoors.getSolution().getConfig();

// get the collisionHandling enum from the config
MPCollisionHandling collisionHandling = config.getCollisionHandling();

// modify the config, this does not have any effect on the SDK yet
config.setEnableClustering(true);

// re-apply the modified config
MapsIndoors.getSolution().setConfig(config);
```

### Main Display Rule

> In V4, the solution config will ship with a Main Display Rule, which is used to define the display rule "ground truths" for the solution. Fields in this display rule must not be `null`. If a display rule with `null` in any fields is provided to the config, an `IllegalArgumentException` will be thrown.

Reviewer: which is used to define the display rule "ground truths" for the solution.

which is used to define the defaults for location and type display rules, when nothing is declared on the specific display rule

        .setDepartureTime(new Date(System.currentTimeMillis()))
        .setOnRouteResultListener((route, error) -> {
            // You get your route (or error) here!
        })
        .build();

MPDirectionsService directionsService = new MPDirectionsService();
directionsService.setConfig(config);
MPPoint from = new MPPoint(57.039395177203936, 9.939182484455051);
MPPoint to = new MPPoint(57.03238690202058, 9.93220061362637);
directionsService.query(from, to);
```

### Render Route

#### V3

To render a given route in V3, instantiate a `MPDirectionsRenderer` with parameters. Then your IDE should be able to show you the various configurable attributes (various animation settings and styling) as well as setting the route. Alternatively, refer to further documentation. To start the renderer/animation, invoke `initMap()`.

```java
MPDirectionsRenderer directionsRenderer = new MPDirectionsRenderer(this, mMap, mMapControl, null);
directionsRenderer.setPolylineAnimated(true);
directionsRenderer.setAnimated(true);
directionsRenderer.setRoute(route);
runOnUiThread( ()-> {
    directionsRenderer.initMap(true);
    directionsRenderer.setRouteLegIndex(0);
});
```

#### V4

> In V4, this has been simplified somewhat. Given a route, you can instantiate a new `MPDirectionsRenderer`, and set the route using `setRoute()`. Use the `MPDirectionsRenderer` object to navigate through the route (next/previous leg) as well as configure the animation and styling of the route on the map. By default the route is animated and repeating, but this is customizable on the `MPDirectionsRenderer` instance.

Reviewer: somewhat seems redundant.
